File: download.go
package main
import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"sync"
)

func downloadFile(url string, destination string, wg *sync.WaitGroup, ch chan bool) {
	defer wg.Done()

	// 创建输出文件
	file, err := os.Create(destination)
	if err != nil {
		fmt.Printf("Failed to create file: %v\n", err)
		return
	}
	defer file.Close()

	// 发起HTTP GET请求
	response, err := http.Get(url)
	if err != nil {
		fmt.Printf("Failed to download file: %v\n", err)
		return
	}
	defer response.Body.Close()

	// 检查HTTP状态码
	if response.StatusCode != http.StatusOK {
		fmt.Printf("Failed to download file, status code: %d\n", response.StatusCode)
		return
	}

	// 将响应体内容写入文件
	_, err = io.Copy(file, response.Body)
	if err != nil {
		fmt.Printf("Failed to write to file: %v\n", err)
		return
	}

	fmt.Printf("Downloaded file: %s\n", destination)
	<-ch // 从信道中释放一个位置，表示下载完成
}

func main() {
	// 读取文件中的URL
	urlFile := "english.txt"
	urls, err := readURLsFromFile(urlFile)
	if err != nil {
		fmt.Printf("Failed to read URLs from file: %v\n", err)
		return
	}

	// 创建目录用于存储下载的文件
	downloadDir := "downloads"
	err = os.MkdirAll(downloadDir, os.ModePerm)
	if err != nil {
		fmt.Printf("Failed to create download directory: %v\n", err)
		return
	}

	// 设置并发下载的最大数量
	maxConcurrentDownloads := 10
	concurrentDownloads := make(chan bool, maxConcurrentDownloads)

	var wg sync.WaitGroup
	// 启动并发下载任务
	for _, url := range urls {
		wg.Add(1)
		filename := filepath.Base(url)
		destination := filepath.Join(downloadDir, filename)
		concurrentDownloads <- true // 从信道中获取一个位置，表示开始下载
		go downloadFile(url, destination, &wg, concurrentDownloads)
	}

	// 等待所有下载任务完成
	wg.Wait()
}

func readURLsFromFile(filePath string) ([]string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var urls []string
	var line string
	for {
		_, err := fmt.Fscanf(file, "%s\n", &line)
		if err != nil {
			if err == io.EOF {
				break
			}
			return nil, err
		}
		urls = append(urls, line)
	}

	return urls, nil
}

File: download.py
import os
import requests
from concurrent.futures import ThreadPoolExecutor
class FileManager:
    def __init__(self):
        self.session = requests.Session()

    def download_image(self, url, destination, headers=None):
        try:
            response = self.session.get(url, stream=True, headers=headers)
            with open(destination, 'wb') as f:
                for chunk in response.iter_content(chunk_size=1024):
                    if chunk:
                        f.write(chunk)
            print(f"Downloaded {destination}")
        except Exception as e:
            print(f"Failed to download {url}: {str(e)}")

    def download_images(self, image_file, download_dir, headers=None):
        os.makedirs(download_dir, exist_ok=True)
        try:
            with open(image_file, 'r') as file:
                image_list = [line.strip() for line in file if line.strip()]
        except Exception as e:
            print(f"Failed to read image file {image_file}: {str(e)}")
            return

        with ThreadPoolExecutor(max_workers=10) as executor:
            for url in image_list:
                if not url:
                    print("Empty URL detected, skipping.")
                    continue
                filename = url.split('/')[-1]
                destination = os.path.join(download_dir, filename)
                executor.submit(self.download_image, url, destination, headers=headers)

# Example usage
file_manager = FileManager()

# File containing image URLs
image_file = 'english.txt'
download_dir = 'image_downloads'
custom_headers = {'User-Agent': 'Mozilla/5.0'}
file_manager.download_images(image_file, download_dir, headers=custom_headers)

